vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+ *
+ */
+#    include "vpr_context.h"
+
+// declaration of used types;
+class AnalyticPlacer;
+struct t_logical_block_type;
+
+// Cut-spreader, as described in HeAP/SimPL papers
+class CutSpreader {
+  public:
+    /*
+     * @brief: Constructor of CutSpreader
+     *
+     * @param analytic_placer	pointer to the analytic_placer that invokes this instance of CutSpreader
+     * 							passed for CutSpreader to directly access data members in analytic_placer such as
+     * 							blk_locs, blk_info, solve_blks, etc, without re-packaging the data to pass to
+     * 							CutSpreader.
+     *
+     * @param vpr_context		legal placement is returned by modifying vpr_context.mutable_placement(), which
+     * 							annealer then operates on. CutSpreader also gets netlist & device info from
+     * 							vpr_context.clustering().clb_nlist and vpr_context.device().
+     *
+     * @param blk_t				logical_block_type for CutSpreader to legalize. Currently can only legalize one
+     * 							type each time.
+     */
+    CutSpreader(AnalyticPlacer* analytic_placer, t_logical_block_type_ptr blk_t);
+
+    /*
+     * @brief: 	Executes the cut-spreader algorithm described in algorithm overview above.
+     * 			Does not include strict_legalize so placement result is not guaranteed to be legal.
+     * 			Strict_legalize must be run after for legal placement result, and for legal placement to
+     * 			be passed to annealer through vpr_ctx.
+     *
+     * 			Input placement is passed by data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to strict legalizer by modifying blk_locs in analytic_placer
+     */
+    void cutSpread();
+
+    /*
+     * @brief: 	Greedy strict legalize using algorithm described in algorithm overview above.
+     *
+     * 			Input illegal placement from data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to annealer by modifying vpr_ctx.mutable_placement()
+     */
+    void strict_legalize();
+
+  private:
+    // marker for blocks that wasn't solved (blocks of different type, macro members)
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // pointer to analytic_placer to access its data members
+    AnalyticPlacer* ap;
+
+    // block type to legalize
+    t_logical_block_type_ptr blk_type;
+
+    // struct that specifies the region occupied by a macro
+    struct MacroExtent {
+        int x0, y0, x1, y1;
+    };
+
+    // struct describing regions on FPGA to cut_spread
+    struct SpreaderRegion {
+        int id;              // index of regions in regions vector
+        int x0, y0, x1, y1;  // extent of the region
+        int n_blks, n_tiles; // number of blocks and compatible tiles
+        bool overused(float beta) const {
+            // determines whether region is overutilized:
+            // overused = (Occupancy / Capacity) > beta
+            if (n_blks > beta * n_tiles) return true;
```

Comment:
Slightly cleaner code to write 
if ( ...) 
   return true
else 
   return false